### ISSUE TITLE
LogManager.ReconfigureExistingLoggers() intermittently throws InvalidOperationException on Travis/Mono

### ISSUE THREAD

**Issue by jpdillingham** [NLog/NLog#2130]
- Bug

**NLog version**: 4.4.9

**Platform**: Mono 4 

Using Xunit to unit test a couple of utility methods I wrote to enable/disable logging levels at runtime, the methods intermittently fail during the CI build with Travis.  I'm also using AppVeyor and have never seen a build fail due to these tests, so it is pretty safe to assume the issue is localized to mono.

- Please post full exception details (message, stacktrace, inner exceptions)
```
    OpenIIoT.Core.Tests.Utility.DisableLoggingLevel [FAIL]
      System.InvalidOperationException : Collection was modified; enumeration operation may not execute.
      Stack Trace:
          at System.ThrowHelper.ThrowInvalidOperationException (System.ExceptionResource resource) [0x0000b] in <4dc8ec68b0964e099af86e50301f5f3c>:0 
          at System.Collections.Generic.Dictionary`2+ValueCollection+Enumerator[TKey,TValue].MoveNext () [0x00013] in <4dc8ec68b0964e099af86e50301f5f3c>:0 
          at NLog.LogFactory+LoggerCache.GetLoggers () [0x00042] in <5652eb00aa3b4e7581fda66dc8b0564d>:0 
          at NLog.LogFactory+LoggerCache.get_Loggers () [0x00000] in <5652eb00aa3b4e7581fda66dc8b0564d>:0 
          at NLog.LogFactory.ReconfigExistingLoggers () [0x00030] in <5652eb00aa3b4e7581fda66dc8b0564d>:0 
          at NLog.LogManager.ReconfigExistingLoggers () [0x00000] in <5652eb00aa3b4e7581fda66dc8b0564d>:0 
          at OpenIIoT.Core.Utility.EnableLoggingLevel (NLog.LogLevel level) [0x00032] in <b6014db257d840ceb2d4ce06ffd49e27>:0 
```

- Are there any work arrounds? yes
I implemented a retry loop that runs 5 times or until the the reconfig runs without error.

Doing a little leg work, the problem appears to be here (line 1354 of LogFactory.cs):

```c#
            private IEnumerable<Logger> GetLoggers()
            {
                // TODO: Test if loggerCache.Values.ToList<Logger>() can be used for the conversion instead.
                List<Logger> values = new List<Logger>(loggerCache.Count);

                foreach (WeakReference loggerReference in loggerCache.Values)
                {
                    Logger logger = loggerReference.Target as Logger;
                    if (logger != null)
                    {
                        values.Add(logger);
                    }
                }

                return values;
            }
```

Clearly the ```Values``` member of  ```loggerCache``` is being updated during the foreach but I lost track of where that might be.  

A quick probable fix would be to set a variable and iterate over that:

```c#
            private IEnumerable<Logger> GetLoggers()
            {
                // TODO: Test if loggerCache.Values.ToList<Logger>() can be used for the conversion instead.
                List<Logger> values = new List<Logger>(loggerCache.Count);

                List<WeakReference> loggerReferences = new List<WeakReference>(loggerCache.Values.ToList());

                foreach (WeakReference loggerReference in loggerReferences)
                {
                    Logger logger = loggerReference.Target as Logger;
                    if (logger != null)
                    {
                        values.Add(logger);
                    }
                }

                return values;
            }
```

I'll submit a PR shortly since I already have it written.

**Comment by 304NotModified** [comment]
Thanks for the PR!

**Comment by 304NotModified** [comment]
Fixed in 4.4.10 and the release is queued. Will be online in a few hours!

Thanks for the submit!
